## Extract Toronto postal codes from Wikipedia page

### Parse Wikipedia page

First task is to extract from a Wikipedia page, based on a provided url, the complete list of postal codes, boroughs and neighborhoods of the Toronto city, and transform later such HTML object into a Pandas dataframe:

In [1]:
#import required libraries
import pandas as pd
import requests
from bs4 import BeautifulSoup

Let's define the seed url, add some headers to avoid banning while scraping, do a request to the site, and check if there is a 200 response from the site:

In [2]:
# define seed url
url = 'https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M'

# identify table class
table_class = "wikitable sortable jquery-tablesorter"

# define headers to avoid banning
headers = {
    "user-agent": "Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Ubuntu Chromium/71.0.3578.80 Chrome/71.0.3578.80 Safari/537.36"
}

# make a request
response = requests.get(url, headers = headers)

# check if a 200 code is returned
print(response.status_code)

200


Generate a soup object and parse the data inside the table:

In [3]:
# generate a soup object and parse the data
soup = BeautifulSoup(response.text)

# parse data from the html into a beautifulsoup object
soup = BeautifulSoup(response.text, 'html.parser')
toronto_table = soup.find('table', {'class':"wikitable"})
toronto_table

<table class="wikitable sortable">
<tbody><tr>
<th>Postal Code
</th>
<th>Borough
</th>
<th>Neighbourhood
</th></tr>
<tr>
<td>M1A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M2A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3A
</td>
<td>North York
</td>
<td>Parkwoods
</td></tr>
<tr>
<td>M4A
</td>
<td>North York
</td>
<td>Victoria Village
</td></tr>
<tr>
<td>M5A
</td>
<td>Downtown Toronto
</td>
<td>Regent Park, Harbourfront
</td></tr>
<tr>
<td>M6A
</td>
<td>North York
</td>
<td>Lawrence Manor, Lawrence Heights
</td></tr>
<tr>
<td>M7A
</td>
<td>Downtown Toronto
</td>
<td>Queen's Park, Ontario Provincial Government
</td></tr>
<tr>
<td>M8A
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M9A
</td>
<td>Etobicoke
</td>
<td>Islington Avenue, Humber Valley Village
</td></tr>
<tr>
<td>M1B
</td>
<td>Scarborough
</td>
<td>Malvern, Rouge
</td></tr>
<tr>
<td>M2B
</td>
<td>Not assigned
</td>
<td>Not assigned
</td></tr>
<tr>
<td>M3B
</td>
<td

### Convert html object into a dataframe

Convert html response into a pandas dataframe:

In [4]:
# set the options of the dataframe display
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

In [5]:
# read html and covert into a dataframe
df_post = pd.read_html(str(toronto_table))
df_post = pd.DataFrame(df_post[0])
df_post.head()

Postal Code           Borough              Neighbourhood
0         M1A      Not assigned               Not assigned
1         M2A      Not assigned               Not assigned
2         M3A        North York                  Parkwoods
3         M4A        North York           Victoria Village
4         M5A  Downtown Toronto  Regent Park, Harbourfront

### Data cleaning

Rename some columns:

In [6]:
# rename columns for ease
df_post = df_post.rename(columns = {"Postal Code": "PostalCode", "Neighbourhood": "Neighborhood"})
df_post.head()

PostalCode           Borough               Neighborhood
0        M1A      Not assigned               Not assigned
1        M2A      Not assigned               Not assigned
2        M3A        North York                  Parkwoods
3        M4A        North York           Victoria Village
4        M5A  Downtown Toronto  Regent Park, Harbourfront

In [7]:
# request the size of the dataframe
df_post.shape

(180, 3)

Clean the row where no Borough info is available:

In [8]:
# clean table of rows with no valid content
df_post = df_post[df_post.Borough != 'Not assigned']
df_post.head()

PostalCode           Borough                                 Neighborhood
2        M3A        North York                                    Parkwoods
3        M4A        North York                             Victoria Village
4        M5A  Downtown Toronto                    Regent Park, Harbourfront
5        M6A        North York             Lawrence Manor, Lawrence Heights
6        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government

In [9]:
df_post.shape

(103, 3)

## Get latitude and longitude values per postal code

### Extract list of postal codes

Firstly, we need to extract the postal codes values from the dataframe as a list for a later iteration of `pgeocode` library over them to get the geographical position through latitude and longitude:   

In [10]:
postal_codes = df_post['PostalCode'].tolist()

### Use a geocoding library to extract the latitude and longitude values

After trying to make use of Google `Geocoder` library and never getting a valid response we imported the `pgeocode` library and made it work by iterating through above list of postal code values:

In [11]:
# already installed, commented line as new installation process always request for a reset of kernel
# uncomment line and install the library if required
# %pip install pgeocode

Let's install the `pgeocode`library and iterate over the generated list above to obtain the latitude and longitude values for every postal code.

After that, we will generate a dataframe with the outcome:

In [12]:
# import library
import pgeocode

# generate an object based on Canadian geography
nomi = pgeocode.Nominatim('ca')

# create an empty list
geo = []

# iterate over the list of postal codes
for item in postal_codes:
    location = nomi.query_postal_code('{}'.format(item))
    geo.append(location)

# add data from the loop to a dataframe    
df_geo = pd.DataFrame(geo)

# format dataframe
df_geo = df_geo[['postal_code', 'latitude', 'longitude']].reset_index()
df_geo = df_geo.rename(columns = {"postal_code": "PostalCode", "latitude": "Latitude", "longitude": "Longitude"})
df_geo = df_geo[['PostalCode', 'Latitude', 'Longitude']]
df_geo.head()

PostalCode  Latitude  Longitude
0        M3A   43.7545   -79.3300
1        M4A   43.7276   -79.3148
2        M5A   43.6555   -79.3626
3        M6A   43.7223   -79.4504
4        M7A   43.6641   -79.3889

In [13]:
df_geo.shape

(103, 3)

### Use default csv file to obtain latitude and longitude values

Since this library seems to return slightly different values of latitude and longitude than expected and also a NaN is returned in both variables for postal code M7R, we decided to consume the data included in the .csv file.

Thus, let's read the data from the csv and generate a dataframe of coordinates: 

In [14]:
# read csv to dataframe
df_coord = pd.read_csv('csv/geospatial_coordinates.csv')

# rename column for postal codes
df_coord = df_coord.rename(columns = {"Postal Code": "PostalCode"})
df_coord.head()

PostalCode   Latitude  Longitude
0        M1B  43.806686 -79.194353
1        M1C  43.784535 -79.160497
2        M1E  43.763573 -79.188711
3        M1G  43.770992 -79.216917
4        M1H  43.773136 -79.239476

### Merge dataframes

Now let´s generate a new dataframe merging with two original ones:

In [15]:
# let´s merge the 2 dataframes
df_toronto = pd.merge(df_post, df_coord, on='PostalCode')

# inspect new dataframe
df_toronto

PostalCode           Borough  \
0          M3A        North York   
1          M4A        North York   
2          M5A  Downtown Toronto   
3          M6A        North York   
4          M7A  Downtown Toronto   
5          M9A         Etobicoke   
6          M1B       Scarborough   
7          M3B        North York   
8          M4B         East York   
9          M5B  Downtown Toronto   
10         M6B        North York   
11         M9B         Etobicoke   
12         M1C       Scarborough   
13         M3C        North York   
14         M4C         East York   
15         M5C  Downtown Toronto   
16         M6C              York   
17         M9C         Etobicoke   
18         M1E       Scarborough   
19         M4E      East Toronto   
20         M5E  Downtown Toronto   
21         M6E              York   
22         M1G       Scarborough   
23         M4G         East York   
24         M5G  Downtown Toronto   
25         M6G  Downtown Toronto   
26         M1H       Scarborough   
27         M2H        North York   
28         M3H        North York   
29         M4H         East York   
30         M5H  Downtown Toronto   
31         M6H      West Toronto   
32         M1J       Scarborough   
33         M2J        North York   
34         M3J        North York   
35         M4J         East York   
36         M5J  Downtown Toronto   
37         M6J      West Toronto   
38         M1K       Scarborough   
39         M2K        North York   
40         M3K        North York   
41         M4K      East Toronto   
42         M5K  Downtown Toronto   
43         M6K      West Toronto   
44         M1L       Scarborough   
45         M2L        North York   
46         M3L        North York   
47         M4L      East Toronto   
48         M5L  Downtown Toronto   
49         M6L        North York   
50         M9L        North York   
51         M1M       Scarborough   
52         M2M        North York   
53         M3M        North York   
54         M4M      East Toronto   
55         M5M        North York   
56         M6M              York   
57         M9M        North York   
58         M1N       Scarborough   
59         M2N        North York   
60         M3N        North York   
61         M4N   Central Toronto   
62         M5N   Central Toronto   
63         M6N              York   
64         M9N              York   
65         M1P       Scarborough   
66         M2P        North York   
67         M4P   Central Toronto   
68         M5P   Central Toronto   
69         M6P      West Toronto   
70         M9P         Etobicoke   
71         M1R       Scarborough   
72         M2R        North York   
73         M4R   Central Toronto   
74         M5R   Central Toronto   
75         M6R      West Toronto   
76         M7R       Mississauga   
77         M9R         Etobicoke   
78         M1S       Scarborough   
79         M4S   Central Toronto   
80         M5S  Downtown Toronto   
81         M6S      West Toronto   
82         M1T       Scarborough   
83         M4T   Central Toronto   
84         M5T  Downtown Toronto   
85         M1V       Scarborough   
86         M4V   Central Toronto   
87         M5V  Downtown Toronto   
88         M8V         Etobicoke   
89         M9V         Etobicoke   
90         M1W       Scarborough   
91         M4W  Downtown Toronto   
92         M5W  Downtown Toronto   
93         M8W         Etobicoke   
94         M9W         Etobicoke   
95         M1X       Scarborough   
96         M4X  Downtown Toronto   
97         M5X  Downtown Toronto   
98         M8X         Etobicoke   
99         M4Y  Downtown Toronto   
100        M7Y      East Toronto   
101        M8Y         Etobicoke   
102        M8Z         Etobicoke   

                                          Neighborhood   Latitude  Longitude  
0                                            Parkwoods  43.753259 -79.329656  
1                                     Victoria Village  43.725882 -79.315572  
2                     

In [16]:
# measure the size of the final dataframe
df_toronto.shape

(103, 5)

## Analyze and cluster Toronto postal codes 

### Plot all Toronto area postal codes in a map

In [17]:
# map rendering library
import folium 

# library to handle data in a vectorized manner
import numpy as np 

# library to handle JSON files
import json

# tranform JSON file into a pandas dataframe
from pandas import json_normalize 

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

print('Libraries imported')

Libraries imported


First, we are going to reuse pgeocode library to extract general coordinates for Toronto, based on a downtown postal code, here M5S, on which we will center the folium map later on:

In [18]:
# generate an object based on Canadian geography
nomi = pgeocode.Nominatim('ca')

# reuse pgeocode library
pc = "M5S"
location = nomi.query_postal_code(pc)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.6629, -79.3987.


Now let's plot with folium a map of the total original 103 postal codes in the dataframe:

In [19]:
# create map of Toronto area using latitude and longitude values
map_toronto_area = folium.Map(location = [latitude, longitude], zoom_start = 10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_toronto['Latitude'], df_toronto['Longitude'], df_toronto['Borough'], df_toronto['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto_area)  
    
map_toronto_area

### Restrict to Toronto city alone

Let's now restrict and continue the analyses and plotting based on boroughs which include Toronto on their name alone:

In [20]:
# restrict dataset to Toronto boroughs
toronto_data = df_toronto[df_toronto['Borough'].str.contains("Toronto")].reset_index(drop = True)

# display new dataframe head and size
print('the Toronto city has ' + str(toronto_data.shape[0]) + ' postal codes')
toronto_data.head(10)

the Toronto city has 39 postal codes


PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   
5        M5E  Downtown Toronto                                  Berczy Park   
6        M5G  Downtown Toronto                           Central Bay Street   
7        M6G  Downtown Toronto                                     Christie   
8        M5H  Downtown Toronto                     Richmond, Adelaide, King   
9        M6H      West Toronto                 Dufferin, Dovercourt Village   

    Latitude  Longitude  
0  43.654260 -79.360636  
1  43.662301 -79.389494  
2  43.657162 -79.378937  
3  43.651494 -79.375418  
4  43.676357 -79.293031  
5  43.644771 -79.373306  
6  43.657952 -79.387383  
7  43.669542 -79.422564  
8  43.650571 -79.384568  
9  43.669005 -79.442259

### Re-plot map for Toronto city postal codes alone

New let's re-plot the map with the restriction to only include Toronto boroughs, as follows:

In [21]:
# create map of Toronto city using latitude and longitude values
map_toronto = folium.Map(location = [latitude, longitude], zoom_start = 12)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html = True)
    folium.CircleMarker(
        [lat, lng],
        radius = 5,
        popup = label,
        color = 'blue',
        fill = True,
        fill_color = '#3186cc',
        fill_opacity = 0.7,
        parse_html = False).add_to(map_toronto)  
    
map_toronto

### Foursquare parameters

Adding the Foursquare API parameters which will help us later use the API to obtain most remarkable venues per postal code of the Toronto city:

In [22]:
# your Foursquare ID
CLIENT_ID = 'client_id' 

# your Foursquare Secret
CLIENT_SECRET = 'client_secret' 

# your FourSquare Access Token
ACCESS_TOKEN = 'access_token' 

If willing to run this full script, you'll need to include your own credentials here.

Once generated the expected data on venues, I will store all the info in a file which could be used for later reference by the cluster analyses so that I can remove my own credentials from the definitive .ipynb file.

### Explore all postal codes in Toronto city

We need to create a function to repeat retrieve the relevant venues over all the postal codes in Toronto city:

In [23]:
# Foursquare API version
VERSION = '20180605'

# a default Foursquare API limit value
LIMIT = 100 

# set values
RADIUS = 500

# postal code latitude value
lat = toronto_data.loc[:, 'Latitude']

# postal code longitude value
lng = toronto_data.loc[:, 'Longitude'] 

With that function we will also append the diverse elements into a new venues dataframe:

In [24]:
# define a function
def getNearbyVenues(names, latitudes, longitudes, radius = 500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            RADIUS, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['PostalCode', 
                  'PostalCode Latitude', 
                  'PostalCode Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

Generate a dataset of venues per postal code based on application of prior function:

In [25]:
# create a dataset with the venues info for Toronto city postal codes
toronto_venues = getNearbyVenues(names = toronto_data['PostalCode'],
                                   latitudes = toronto_data['Latitude'],
                                   longitudes = toronto_data['Longitude']
                                  )

M5A
M7A
M5B
M5C
M4E
M5E
M5G
M6G
M5H
M6H
M5J
M6J
M4K
M5K
M6K
M4L
M5L
M4M
M4N
M5N
M4P
M5P
M6P
M4R
M5R
M6R
M4S
M5S
M6S
M4T
M5T
M4V
M5V
M4W
M5W
M4X
M5X
M4Y
M7Y


Check the size of the dataframe and explore the first lines:

In [26]:
# print size of dataframe
print(toronto_venues.shape)

# display sample of the dataframe 
toronto_venues.head()

(1604, 7)


PostalCode  PostalCode Latitude  PostalCode Longitude  \
0        M5A             43.65426            -79.360636   
1        M5A             43.65426            -79.360636   
2        M5A             43.65426            -79.360636   
3        M5A             43.65426            -79.360636   
4        M5A             43.65426            -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

As commented above, let's save the dataframe into a .csv file from which to work on: 

In [27]:
# save dataframe into .csv
toronto_venues.to_csv('csv/toronto_venues.csv')

### Review venues

Let's first import the info in the stored .csv file into a dataframe:

In [28]:
toronto_venues = pd.read_csv('csv/toronto_venues.csv', index_col = None).drop('Unnamed: 0', axis=1)
toronto_venues.head()

PostalCode  PostalCode Latitude  PostalCode Longitude  \
0        M5A             43.65426            -79.360636   
1        M5A             43.65426            -79.360636   
2        M5A             43.65426            -79.360636   
3        M5A             43.65426            -79.360636   
4        M5A             43.65426            -79.360636   

                    Venue  Venue Latitude  Venue Longitude  \
0        Roselle Desserts       43.653447       -79.362017   
1           Tandem Coffee       43.653559       -79.361809   
2  Cooper Koo Family YMCA       43.653249       -79.358008   
3     Body Blitz Spa East       43.654735       -79.359874   
4          Impact Kitchen       43.656369       -79.356980   

        Venue Category  
0               Bakery  
1          Coffee Shop  
2  Distribution Center  
3                  Spa  
4           Restaurant

Let's check how many venues were returned for each postal code:

In [29]:
# group and count returned venues by postal code
toronto_venues.groupby('PostalCode').count()['Venue']

PostalCode
M4E      5
M4K     43
M4L     20
M4M     36
M4N      4
M4P      8
M4R     17
M4S     34
M4T      3
M4V     16
M4W      4
M4X     45
M4Y     75
M5A     46
M5B    100
M5C     81
M5E     59
M5G     64
M5H     94
M5J    100
M5K    100
M5L    100
M5N      2
M5P      4
M5R     20
M5S     32
M5T     63
M5V     15
M5W     99
M5X    100
M6G     16
M6H     17
M6J     42
M6K     22
M6P     25
M6R     14
M6S     33
M7A     30
M7Y     16
Name: Venue, dtype: int64

Let's find out how many unique categories can be curated from all the returned venues:

In [30]:
# extract unique categories for venues
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 235 uniques categories.


### Analyze each postal code based on venues included

Let's do now a `one-hot` encoding of the diverse venues by postal code and generate a table with split categories:

In [31]:
# one-hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add postal code column back to dataframe
toronto_onehot['PostalCode'] = toronto_venues['PostalCode'] 

# move postal code column to the first column
fixed_columns = [toronto_onehot.columns[-1]] + list(toronto_onehot.columns[:-1])
toronto_onehot = toronto_onehot[fixed_columns]

toronto_onehot.head()

PostalCode  Airport  Airport Food Court  Airport Lounge  Airport Service  \
0        M5A        0                   0               0                0   
1        M5A        0                   0               0                0   
2        M5A        0                   0               0                0   
3        M5A        0                   0               0                0   
4        M5A        0                   0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  Art Gallery  \
0                 0                    0             0         0            0   
1                 0                    0             0         0            0   
2                 0                    0             0         0            0   
3                 0                    0             0         0            0   
4                 0                    0             0         0            0   

   Art Museum  Arts & Crafts Store  Asian Restaurant  Athletics & Sports  \
0           0                    0                 0                   0   
1           0                    0                 0                   0   
2           0                    0                 0                   0   
3           0                    0                 0                   0   
4           0                    0                 0                   0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       1     0    0   
1              0          0           0           0       0     0    0   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Stadium  Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  \
0                 0                   0      0                0         0   
1                 0                   0      0                0         0   
2                 0                   0      0                0         0   
3                 0                   0      0                0         0   
4                 0                   0      0                0         0   

   Beer Store  Belgian Restaurant  Bike Rental / Bike Share  Bistro  \
0           0                   0                         0       0   
1           0                   0                         0       0   
2           0                   0                         0       0   
3           0                   0                         0       0   
4           0                   0                         0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Burger Joint  Burrito Place  Bus Line  \
0        0                0             0              0         0   
1        0                0             0              0         0   
2        0                0             0              0         0   
3        0                0             0              0         0   
4        0                0             0              0         0   

   Business Service  Butcher  Café  Cajun / Creole Restaurant  Candy Store  \
0                 0        0     0                          0            0   
1                 0        0     0                          0            0   
2                 0        0     0                          0            0   
3                 0        0     0            

In [32]:
# new dataframe size
toronto_onehot.shape

(1604, 236)

Next, let's group rows by postal code and by taking the mean of the frequency of occurrence of each category:

In [33]:
# generate a new dataframe based on grouped data by postal code
toronto_grouped = toronto_onehot.groupby('PostalCode').mean().reset_index()
toronto_grouped

PostalCode   Airport  Airport Food Court  Airport Lounge  Airport Service  \
0         M4E  0.000000            0.000000        0.000000              0.0   
1         M4K  0.000000            0.000000        0.000000              0.0   
2         M4L  0.000000            0.000000        0.000000              0.0   
3         M4M  0.000000            0.000000        0.000000              0.0   
4         M4N  0.000000            0.000000        0.000000              0.0   
5         M4P  0.000000            0.000000        0.000000              0.0   
6         M4R  0.000000            0.000000        0.000000              0.0   
7         M4S  0.000000            0.000000        0.000000              0.0   
8         M4T  0.000000            0.000000        0.000000              0.0   
9         M4V  0.000000            0.000000        0.000000              0.0   
10        M4W  0.000000            0.000000        0.000000              0.0   
11        M4X  0.000000            0.000000        0.000000              0.0   
12        M4Y  0.000000            0.000000        0.000000              0.0   
13        M5A  0.000000            0.000000        0.000000              0.0   
14        M5B  0.000000            0.000000        0.000000              0.0   
15        M5C  0.000000            0.000000        0.000000              0.0   
16        M5E  0.000000            0.000000        0.000000              0.0   
17        M5G  0.000000            0.000000        0.000000              0.0   
18        M5H  0.000000            0.000000        0.000000              0.0   
19        M5J  0.000000            0.000000        0.000000              0.0   
20        M5K  0.000000            0.000000        0.000000              0.0   
21        M5L  0.000000            0.000000        0.000000              0.0   
22        M5N  0.000000            0.000000        0.000000              0.0   
23        M5P  0.000000            0.000000        0.000000              0.0   
24        M5R  0.000000            0.000000        0.000000              0.0   
25        M5S  0.000000            0.000000        0.000000              0.0   
26        M5T  0.000000            0.000000        0.000000              0.0   
27        M5V  0.066667            0.066667        0.133333              0.2   
28        M5W  0.000000            0.000000        0.000000              0.0   
29        M5X  0.000000            0.000000        0.000000              0.0   
30        M6G  0.000000            0.000000        0.000000              0.0   
31        M6H  0.000000            0.000000        0.000000              0.0   
32        M6J  0.000000            0.000000        0.000000              0.0   
33        M6K  0.000000            0.000000        0.000000              0.0   
34        M6P  0.000000            0.000000        0.000000              0.0   
35        M6R  0.000000            0.000000        0.000000              0.0   
36        M6S  0.000000            0.000000        0.000000              0.0   
37        M7A  0.000000            0.000000        0.000000              0.0   
38        M7Y  0.000000            0.000000        0.000000              0.0   

    Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0           0.000000             0.000000      0.000000      0.00   
1           0.000000             0.023256      0.000000      0.00   
2           0.000000             0.000000      0.000000      0.00   
3           0.000000             0.055556      0.000000      0.00   
4           0.000000             0.000000      0.000000      0.00   
5           0.000000             0.000000      0.000000      0.00   
6           0.000000             0.000000      0.000000      0.00   
7           0.000000             0.000000      0.000000      0.00   
8           0.000000             0.000000      0.000000      0.00   
9           0.000000             0.062500      0.000000      0.00   
10          0.000000             0.000000  

In [34]:
# let's confirm the new size
toronto_grouped.shape

(39, 236)

Let's write a function to sort the venues in descending order:

In [35]:
# define a function to sort the venues in descending order
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending = False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each postal code:

In [36]:
# define the number of top venues
num_top_venues = 10

# include some position indicators
indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['PostalCode']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
postal_code_venues_sorted = pd.DataFrame(columns = columns)
postal_code_venues_sorted['PostalCode'] = toronto_grouped['PostalCode']

for ind in np.arange(toronto_grouped.shape[0]):
    postal_code_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

postal_code_venues_sorted.head()

PostalCode 1st Most Common Venue 2nd Most Common Venue  \
0        M4E          Neighborhood     Health Food Store   
1        M4K      Greek Restaurant           Coffee Shop   
2        M4L                  Park  Fast Food Restaurant   
3        M4M           Coffee Shop   American Restaurant   
4        M4N                  Park              Bus Line   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0      Asian Restaurant                 Trail                     Pub   
1    Italian Restaurant        Ice Cream Shop  Furniture / Home Store   
2      Sushi Restaurant         Movie Theater          Sandwich Place   
3                Bakery               Brewery                    Café   
4      Business Service           Swim School             Yoga Studio   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0           Yoga Studio      Doner Restaurant        Discount Store   
1           Yoga Studio       Bubble Tea Shop                   Spa   
2            Restaurant    Italian Restaurant     Fish & Chips Shop   
3             Gastropub           Yoga Studio           Fish Market   
4        Discount Store  Ethiopian Restaurant           Escape Room   

  9th Most Common Venue       10th Most Common Venue  
0   Distribution Center                      Dog Run  
1   Japanese Restaurant               Cosmetics Shop  
2          Intersection                   Steakhouse  
3                  Park                 Neighborhood  
4     Electronics Store  Eastern European Restaurant

### Cluster postal codes

Run k-means to cluster the neighborhood into 5 clusters:

In [37]:
# set number of clusters
kclusters = 7

toronto_grouped_clustering = toronto_grouped.drop('PostalCode', 1)

# run k-means clustering
kmeans = KMeans(n_clusters = kclusters, random_state = 0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([4, 0, 0, 0, 2, 0, 0, 0, 3, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each postal code:

In [38]:
# add clustering labels
postal_code_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each postal code
toronto_merged = toronto_merged.join(postal_code_venues_sorted.set_index('PostalCode'), on = 'PostalCode')

# check the last columns
toronto_merged.head() 

PostalCode           Borough                                 Neighborhood  \
0        M5A  Downtown Toronto                    Regent Park, Harbourfront   
1        M7A  Downtown Toronto  Queen's Park, Ontario Provincial Government   
2        M5B  Downtown Toronto                     Garden District, Ryerson   
3        M5C  Downtown Toronto                               St. James Town   
4        M4E      East Toronto                                  The Beaches   

    Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
0  43.654260 -79.360636               0           Coffee Shop   
1  43.662301 -79.389494               0           Coffee Shop   
2  43.657162 -79.378937               0           Coffee Shop   
3  43.651494 -79.375418               0           Coffee Shop   
4  43.676357 -79.293031               4          Neighborhood   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Park                Bakery        Breakfast Spot   
1      Sushi Restaurant                 Diner           Yoga Studio   
2        Clothing Store    Italian Restaurant        Cosmetics Shop   
3                  Café          Cocktail Bar   American Restaurant   
4     Health Food Store      Asian Restaurant                 Trail   

  5th Most Common Venue 6th Most Common Venue      7th Most Common Venue  \
0                  Café                   Pub                    Theater   
1    College Auditorium              Beer Bar              Smoothie Shop   
2                 Hotel       Bubble Tea Shop  Middle Eastern Restaurant   
3             Gastropub   Moroccan Restaurant           Department Store   
4                   Pub           Yoga Studio           Doner Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0     French Restaurant      Greek Restaurant              Wine Shop  
1        Sandwich Place         Burrito Place             Restaurant  
2                  Café   Japanese Restaurant         Lingerie Store  
3        Lingerie Store                   Gym                   Park  
4        Discount Store   Distribution Center                Dog Run

Finally, let's visualize the resulting clusters:

In [39]:
# create map
map_clusters = folium.Map(location = [latitude, longitude], zoom_start = 12)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html = True)
    folium.CircleMarker(
        [lat, lon],
        radius = 5,
        popup = label,
        color = rainbow[cluster-1],
        fill = True,
        fill_color = rainbow[cluster-1],
        fill_opacity = 0.7).add_to(map_clusters)
       
map_clusters

### Analyze the clusters

We can examine each cluster and determine the discriminating venue categories that distinguish each cluster. Based on the defining categories, we can then assign a name to each cluster:

In [40]:
# cluster 1
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
0   Downtown Toronto               0           Coffee Shop   
1   Downtown Toronto               0           Coffee Shop   
2   Downtown Toronto               0           Coffee Shop   
3   Downtown Toronto               0           Coffee Shop   
5   Downtown Toronto               0           Coffee Shop   
6   Downtown Toronto               0           Coffee Shop   
7   Downtown Toronto               0         Grocery Store   
8   Downtown Toronto               0           Coffee Shop   
9       West Toronto               0                Bakery   
10  Downtown Toronto               0           Coffee Shop   
11      West Toronto               0                   Bar   
12      East Toronto               0      Greek Restaurant   
13  Downtown Toronto               0           Coffee Shop   
14      West Toronto               0                  Café   
15      East Toronto               0                  Park   
16  Downtown Toronto               0           Coffee Shop   
17      East Toronto               0           Coffee Shop   
20   Central Toronto               0  Gym / Fitness Center   
22      West Toronto               0       Thai Restaurant   
23   Central Toronto               0        Clothing Store   
24   Central Toronto               0        Sandwich Place   
25      West Toronto               0        Breakfast Spot   
26   Central Toronto               0          Dessert Shop   
27  Downtown Toronto               0                  Café   
28      West Toronto               0                  Café   
30  Downtown Toronto               0                  Café   
31   Central Toronto               0           Coffee Shop   
32  Downtown Toronto               0       Airport Service   
34  Downtown Toronto               0           Coffee Shop   
35  Downtown Toronto               0                  Café   
36  Downtown Toronto               0           Coffee Shop   
37  Downtown Toronto               0           Coffee Shop   
38      East Toronto               0    Light Rail Station   

   2nd Most Common Venue  3rd Most Common Venue  \
0                   Park                 Bakery   
1       Sushi Restaurant                  Diner   
2         Clothing Store     Italian Restaurant   
3                   Café           Cocktail Bar   
5           Cocktail Bar                 Bakery   
6         Sandwich Place     Italian Restaurant   
7                   Café                   Park   
8                   Café             Restaurant   
9               Pharmacy            Music Venue   
10              Aquarium                   Café   
11            Restaurant                   Café   
12           Coffee Shop     Italian Restaurant   
13                 Hotel                   Café   
14           Coffee Shop         Breakfast Spot   
15  Fast Food Restaurant       Sushi Restaurant   
16            Restaurant                   Café   
17   American Restaurant                 Bakery   
20        Breakfast Spot                  Hotel   
22                  Café     Mexican Restaurant   
23           Coffee Shop            Yoga Studio   
24                  Café            Coffee Shop   
25             Gift Shop           Dessert Shop   
26           Pizza Place         Sandwich Place   
27                   Bar    Japanese Restaurant   
28           Coffee Shop     Italian Restaurant   
30           Coffee Shop  Vietnamese Restaurant   
31    Light Rail Station             Bagel Shop   
32        Airport Lounge                Airport   
34    Seafood Restaurant             Restaurant   
35           Coffee Shop             Restaurant   
36                  Café                  Hotel   
37      Sushi Restaurant    Japanese Restaurant   
38           Yoga Studio          Auto Workshop   

            4th Most Common Venue   5th Most Common Venue  \
0                  Breakfast Spot                    Café   
1                     Yoga Studio      College 

In [41]:
# cluster 2
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels    1st Most Common Venue  \
19  Central Toronto               1  Health & Beauty Service   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
19                Garden           Yoga Studio          Dessert Shop   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
19           Event Space  Ethiopian Restaurant           Escape Room   

   8th Most Common Venue        9th Most Common Venue 10th Most Common Venue  
19     Electronics Store  Eastern European Restaurant    Dumpling Restaurant

In [42]:
# cluster 3
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
18  Central Toronto               2                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
18              Bus Line      Business Service           Swim School   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
18           Yoga Studio        Discount Store  Ethiopian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
18           Escape Room     Electronics Store  Eastern European Restaurant

In [43]:
# cluster 4
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue  \
29  Central Toronto               3                Lawyer   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
29                 Trail           Summer Camp           Yoga Studio   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
29                 Diner           Event Space  Ethiopian Restaurant   

   8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
29           Escape Room     Electronics Store  Eastern European Restaurant

In [44]:
# cluster 5
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]

Borough  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
4  East Toronto               4          Neighborhood     Health Food Store   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
4      Asian Restaurant                 Trail                   Pub   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
4           Yoga Studio      Doner Restaurant        Discount Store   

  9th Most Common Venue 10th Most Common Venue  
4   Distribution Center                Dog Run

Based on data defined in the clusters we could define that the postal codes included in each are representing the following activities and space utilization: 

**cluster 1:** crazy for the coffee (coffee shops, bakeries, breakfasts, etc.)

**cluster 2:** take care of yourself (beauty, yoga, desserts, etc.)

**cluster 3:** stress out (park, swimming pool, yoga, etc.)

**cluster 4:** let's go professional (lawyer, diner, event space, etc.)

**cluster 5:** living area (neighborhood, health food store, dog run, etc.)

Hope you enjoyed the journey across the main areas of Toronto city as much as I did!

**Author: Sergio Hidalgo - March 2021**